politeiad/backendv2/tstorebe/plugins/pi/hooks.go

```diff
@@ -390,11 +395,246 @@ func (p *piPlugin) voteSummary(token []byte) (*ticketvote.SummaryReply, error) {
 	return &sr, nil
 }
 
+// comments requests all comments on a record from the comments plugin.
+func (p *piPlugin) comments(token []byte) (*comments.GetAllReply, error) {
+	reply, err := p.backend.PluginRead(token, comments.PluginID,
+		comments.CmdGetAll, "")
+	if err != nil {
+		return nil, err
+	}
+	var gar comments.GetAllReply
+	err = json.Unmarshal([]byte(reply), &gar)
+	if err != nil {
+		return nil, err
+	}
+	return &gar, nil
+}
+
+// isAncestorOf returns whether the given parentID is an ancestor of the given
+// childID comment, by traveling the comment tree defined by
+// the ParentID of the child comment.
+func (p *piPlugin) isAncestorOf(ancestorID, childID uint32, cs []comments.Comment) bool {
+	if ancestorID == childID {
+		return true
+	}
+	// Convert comments slice to a map
+	commentsMap := make(map[uint32]comments.Comment, len(cs))
+	for _, c := range cs {
+		commentsMap[c.CommentID] = c
+	}
+	// Travel the comment tree to search for an ancestor
+	for {
+		current := commentsMap[childID]
+		// If we reach the tree haed without visiting the provided
+		// ancestorID then it is not an ancestor.
+		if current.ParentID == 0 {
+			break
+		}
+		// Check if next parent in the tree is the ancestor
+		if current.ParentID == ancestorID {
+			return true
+		}
+		childID = current.ParentID
+	}
+	return false
+}
+
+// latestAuthorUpdate gets the latest author update on a record, if
+// the record has no author update it returns nil.
+func (p *piPlugin) latestAuthorUpdate(token []byte, cs []comments.Comment) (*comments.Comment, error) {
+	var latestAuthorUpdate comments.Comment
+	for _, c := range cs {
+		if c.ExtraDataHint != pi.ProposalUpdateHint {
+			continue
+		}
+		if c.Timestamp > latestAuthorUpdate.Timestamp {
+			latestAuthorUpdate = c
+		}
+	}
+	return &latestAuthorUpdate, nil
+}
+
+// recordAuthor returns the author's userID of the record associated with
+// the provided token.
+func (p *piPlugin) recordAuthor(token []byte) (string, error) {
+	reply, err := p.backend.PluginRead(token, usermd.PluginID,
+		usermd.CmdAuthor, "")
+	if err != nil {
+		return "", err
+	}
+	var ar usermd.AuthorReply
+	err = json.Unmarshal([]byte(reply), &ar)
+	if err != nil {
+		return "", err
+	}
+	return ar.UserID, nil
+}
+
+// commentVoteAllowedOnApprovedProposal verifies that the given comment
+// vote is allowed on the approved proposal associated with the provided
+// token.
+func (p *piPlugin) commentVoteAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var v comments.Vote
+	err := json.Unmarshal([]byte(payload), &v)
+	if err != nil {
+		return err
+	}
+	if !p.isAncestorOf(latestAuthorUpdate.CommentID, v.CommentID,
+		cs) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeVoteStatusInvalid),
```

Comment:
This doesn't seem like the appropriate error.  The vote status allows for writes.  The issue is that the writes are only allowed on updates.